middle_end/flambda/types/inlining/inlining_arguments.ml

```diff
@@ -14,65 +14,158 @@
+    && Float.compare t1.prim_cost t2.prim_cost = 0
+    && Float.compare t1.branch_cost t2.branch_cost = 0
+    && Float.compare t1.indirect_cost t2.indirect_cost = 0
+    && t1.small_function_size = t2.small_function_size
+    && t1.large_function_size = t2.large_function_size
+    && Float.compare t1.threshold t2.threshold = 0
+
+  let (<=) t1 t2 =
+    t1.max_inlining_depth <= t2.max_inlining_depth
+    && Float.compare t1.call_cost t2.call_cost <= 0
+    && Float.compare t1.alloc_cost t2.alloc_cost <= 0
+    && Float.compare t1.prim_cost t2.prim_cost <= 0
+    && Float.compare t1.branch_cost t2.branch_cost <= 0
+    && Float.compare t1.indirect_cost t2.indirect_cost <= 0
+
+  let meet args1 args2 =
+    {
+      max_inlining_depth = min args1.max_inlining_depth args2.max_inlining_depth;
+      call_cost = Float.min args1.call_cost args2.call_cost;
+      alloc_cost = Float.min args1.alloc_cost args2.alloc_cost;
+      prim_cost = Float.min args1.prim_cost args2.prim_cost;
+      branch_cost = Float.min args1.branch_cost args2.branch_cost;
+      indirect_cost = Float.min args1.indirect_cost args2.indirect_cost;
+      small_function_size = min args1.small_function_size args2.small_function_size;
+      large_function_size = min args1.large_function_size args2.large_function_size;
+      threshold = Float.min args1.threshold args2.threshold
+    }
+
+
+  let cost_i (flag : Clflags.Int_arg_helper.parsed) ~round =
+    Clflags.Int_arg_helper.get ~key:round flag
+
+  let cost_f (flag : Clflags.Float_arg_helper.parsed) ~round =
+    Clflags.Float_arg_helper.get ~key:round flag
+
+  let create ~round = {
+    max_inlining_depth = !Clflags.Flambda.Expert.max_inlining_depth;
+    call_cost = cost_f !Clflags.inline_call_cost ~round;
+    alloc_cost = cost_f !Clflags.inline_alloc_cost ~round;
+    prim_cost = cost_f !Clflags.inline_prim_cost ~round;
+    branch_cost = cost_f !Clflags.inline_branch_cost ~round;
+    indirect_cost = cost_f !Clflags.inline_indirect_cost ~round;
+    small_function_size = cost_i !Clflags.inline_small_function_size ~round;
+    large_function_size = cost_i !Clflags.inline_large_function_size ~round;
+    threshold = cost_f !Clflags.inline_threshold ~round;
+  }
+end
+
+type t = Args.t Or_unknown.t
+
+let unknown = Or_unknown.Unknown
+
+let print ppf = Or_unknown.print Args.print ppf
 
 let max_inlining_depth = function
-  | Unknown -> 0
-  | O1 -> 1
-  | O2 -> 2
-  | O3 -> 3
-  | Custom t -> t.max_inlining_depth
-
-let equal t1 t2 =
-  max_inlining_depth t1 = max_inlining_depth t2
-
-let compress t =
-  match t with
-  | O1 | O2 | O3 | Unknown -> t
-  | t ->
-    if equal t O1 then O1
-    else if equal t O2 then O2
-    else if equal t O3 then O3
-    else t
-
-let meet_custom t1 t2 =
-  let max_inlining_depth =
-    if (max_inlining_depth t1) <= (max_inlining_depth t2) then
-      max_inlining_depth t1
-    else
-      max_inlining_depth t2
-  in
-  Custom { max_inlining_depth }
+  | Or_unknown.Unknown -> assert false
```

Comment:
This should use `Misc.fatal_errorf` rather than `assert false`, since this case is not obviously impossible just looking locally at the code nearby in this file.  In fact maybe there should be a single wrapper function that checks for `Unknown`, then the failure case can be factored out into there.

Reply:
Fixed